src/serializers/fields.rs

```diff
@@ -120,24 +120,21 @@ impl GeneralFieldsSerializer {
 
     fn extract_dicts<'a>(&self, value: &'a PyAny) -> Option<(&'a PyDict, Option<&'a PyDict>)> {
         match self.mode {
-            FieldsMode::ModelExtra => {
-                if let Ok((main_dict, extra_dict)) = value.extract::<(&PyDict, &PyDict)>() {
-                    Some((main_dict, Some(extra_dict)))
-                } else {
-                    None
-                }
-            }
-            _ => {
-                if let Ok(main_dict) = value.downcast::<PyDict>() {
-                    Some((main_dict, None))
-                } else {
-                    None
-                }
-            }
+            FieldsMode::ModelExtra => value
+                .extract::<(&PyDict, &PyDict)>()
+                .ok()
+                .map(|(main_dict, extra_dict)| (main_dict, Some(extra_dict))),
+            _ => value.downcast::<PyDict>().ok().map(|main_dict| (main_dict, None)),
         }
     }
 }
 
+/// Collect all the items in a dict into a vector.
+/// This is to allow safe iteration while the dict might be modified.
+fn dict_items(dict: &PyDict) -> Vec<(&PyAny, &PyAny)> {
```

Comment:
Potentially worth trying a `SmallVec` here, maybe 4, 6 or 8 fields, assuming that most models are small then that'll help avoid heap allocations.

Reply:
I'd really rather we didn't allocate like this. Can we not implement our own fallible version of dict iter and fixed the cached properties version like this?

Otherwise we need:
* some benchmarks to check the affect of this, I've no idea if our current benchmarks cover this case
* `SmallVec` with size 16 or something which I would guess covers 90+% of models